### DISCUSSION TITLE
Update in canopy_server/app.py is not reflecting the update and using default version

### DISCUSSION THREAD

**Discussion by tulaseerao** [pinecone-io/canopy#275] [Q&A]
Hello folks,
I tried to update upsert route(/upsert) in canopy_server/app.py to add new parameter called index_name to override the index name for upsertion. I made all the changes to app.py and tried it locally and still using default behavior. Is there anything I need to delete and restart the server to reflect my changes?

Here is the snippet I'm trying to change

```
upsert_index_name = request.index_name or os.getenv("INDEX_NAME")
        logger.info(f"upsert_index_name: {upsert_index_name}")
        kb_upsert = KnowledgeBase(index_name=upsert_index_name)
        kb_upsert.connect()
```

Over all code
`async def upsert(
    request: ContextUpsertRequest = Body(...),
) -> SuccessUpsertResponse:
    """
    Upsert documents into the knowledge base. Upserting is a way to add new documents or update existing ones.
    Each document has a unique ID. If a document with the same ID already exists, it is updated.

    The documents are chunked and encoded, then the resulting encoded chunks are sent to the Pinecone index in batches.
    """  # noqa: E501
    try:
        logger.info(f"Upserting {len(request.documents)} documents")
        upsert_index_name = request.index_name or os.getenv("INDEX_NAME")
        logger.info(f"upsert_index_name: {upsert_index_name}")
        kb_upsert = KnowledgeBase(index_name=upsert_index_name)
        kb_upsert.connect()
        await run_in_threadpool(
            kb_upsert.upsert, documents=request.documents, batch_size=request.batch_size
        )

        return SuccessUpsertResponse()

    except Exception as e:
        logger.exception(e)
        raise HTTPException(status_code=500, detail=f"Internal Service Error: {str(e)}")
`

**Comment by igiloh-pinecone** [comment]
Hi @tulaseerao, all in all your code seems correct.  
If your canopy server was already running when you changed the code, and you haven't run it initially with the `--reload` flag - then yes, you will need to one time kill the server and start it anew for the code changes to take place.  
From that point on, no further restarts should be required. If for some reason this is still not working, write back here and we'll try to assist.

**Reply by tulaseerao** [reply]
Hello @igiloh-pinecone,

It seems I'm having a bit of a puzzle with the server (`canopy start`). Although I've terminated and restarted it, the server appears to still pull from the canopy-sdk or some sort of cache somewhere.

The slightly baffling part is that even after I removed the entire canopy-server directory and fired up the server, it seems to function as normal. This led me to believe there could be an issue in my local environment. To test it out, I deployed the same code on Heroku, but, even there, my code updates do not seem to be reflecting.

Any insights or advice to resolve this would be appreciated.

**Reply by igiloh-pinecone** [reply]
Hi @tulaseerao when you run the command `canopy start`, it simply run the version of Canopy installed in your current machine \ virtual env.  
If you installed canopy with `pip install canopy-sdk`, and later separately cloned the repo, from Python's perspective the version to use is the installed ones.  
Can you please run these two commands, and see what is the output:
```bash
> which canopy
> python -c "import canopy; print(canopy)"
```
If the paths you see are not where you've edited the code, but instead where Canopy was installed by `pip` - then the behavior you're describing makes perfect sense.

If you want to run Canopy from source, after making some local code changes - please see the [building from source](https://github.com/pinecone-io/canopy/blob/main/CONTRIBUTING.md#building-from-source) instructions.

**Reply by tulaseerao** [reply]
Thanks @igiloh-pinecone. This helped and I uninstalled pinecone-sdk in my local and it worked fine.  Thanks for your help
